smartsim/entity/ensemble.py

```diff
@@ -59,23 +59,182 @@ def __init__(
         replicas: int = 1,
     ) -> None:
+        """Initialize an ``Ensemble`` of application instances
+
+        :param name: name of the ensemble
+        :param exe: executable to run
+        :param exe_args: executable arguments
+        :param exe_arg_parameters: parameters and values to be used when configuring entities
+        :param files: files to be copied, symlinked, and/or configured prior to
+                      execution
+        :param file_parameters: parameters and values to be used when configuring
+                                files
+        :param permutation_strategy: strategy to control how the param values are applied to the Ensemble
+        :param max_permutations: max parameter permutations to set for the ensemble
+        :param replicas: number of identical entities to create within an Ensemble
+        :return: and ``Ensemble`` instance
+        """
         self.name = name
-        self.exe = os.fspath(exe)
-        self.exe_args = list(exe_args) if exe_args else []
-        self.exe_arg_parameters = (
+        """The name of the ensemble """
+        self._exe = os.fspath(exe)
+        """The executable to run """
+        self._exe_args = list(exe_args) if exe_args else []
+        """The executable arguments """
+        self._exe_arg_parameters = (
             copy.deepcopy(exe_arg_parameters) if exe_arg_parameters else {}
         )
-        self.files = copy.deepcopy(files) if files else EntityFiles()
-        self.file_parameters = dict(file_parameters) if file_parameters else {}
-        self.permutation_strategy = permutation_strategy
-        self.max_permutations = max_permutations
-        self.replicas = replicas
+        """The parameters and values to be used when configuring entities"""
+        self._files = copy.deepcopy(files) if files else EntityFiles()
+        """The files to be copied, symlinked, and/or configured prior to execution"""
+        self._file_parameters = dict(file_parameters) if file_parameters else {}
+        """The parameters and values to be used when configuring files"""
+        self._permutation_strategy = permutation_strategy
+        """The strategy to control how the param values are applied to the Ensemble"""
+        self._max_permutations = max_permutations
+        """The maximum number of entities to come out of the permutation strategy"""
+        self._replicas = replicas
+        """How many identical entities to create within an Ensemble"""
+
+    @property
+    def exe(self) -> str:
+        """Return executable to run.
+
+        :returns: application executable to run
+        """
+        return self._exe
+
+    @exe.setter
+    def exe(self, value: str | os.PathLike[str]) -> None:
+        """Set executable to run.
+
+        :param value: executable to run
+        """
+        self._exe = os.fspath(value)
+
+    @property
+    def exe_args(self) -> t.Sequence[str]:
+        """Return an immutable list of attached executable arguments.
```

Comment:
Forget how we solved this with `Application` but I think we didn't like the "typed as 'immutable + Sequence' when in reality the return type is a `list[str]`" with its `exe_args`. Can you revise to match?

Reply:
changed